Credit provider integration Python API (Part 2 of 3)

wedaly opened the pull request:
The exciting sequel to https://github.com/edx/edx-platform/pull/8338!

* Add end-point for initiating a request for credit from a provider.
* Add an end-point for a provider to update the status of a request (approved / denied).

JIRA: [ECOM-1693](https://openedx.atlassian.net/browse/ECOM-1693)

@clintonb please review.
FYI: @aamir-khan @zubair-arbi @awais786 @ahsan-ul-haq

wedaly commented on openedx/core/djangoapps/credit/views.py:
```diff
@@ -0,0 +1,292 @@
+"""
```
Note to reviewers: throughout this file, I have chosen not to internationalize the error messages returned by the API.  I expect these errors to be used for debugging purposes (especially for external credit providers integrating with us) and not shown to end-users.

wedaly commented:
Tests are failing in Jenkins but passing locally, most likely related to feature-flagging the URLs.  Will investigate and fix today.

wedaly commented:
Okay, I think I figured out the Jenkins test failure issue.  Updated the PR to reload the URLs correctly.

wedaly commented:
jenkins test this please

clintonb commented on openedx/core/djangoapps/credit/urls.py:
```diff
@@ -0,0 +1,26 @@
+
+urlpatterns = []
+
+if settings.FEATURES.get("ENABLE_CREDIT_API"):
```
Move this to `lms/urls.py` so that future developers don't have to dig too deep into the code to find it.

clintonb commented on openedx/core/djangoapps/credit/api.py:
```diff
@@ -191,8 +210,36 @@ def create_credit_request(course_key, provider_id, username):
     except CreditEligibility.DoesNotExist:
+        log.warning(
+            u'User tried to initiate a request for credit, '
+            u'but the user is not eligible for credit'
```
This can be on a single line.

clintonb commented on openedx/core/djangoapps/credit/exceptions.py:
```diff
@@ -1,42 +1,57 @@
+class CreditApiBadRequest(Exception):
+    """
+    Could not complete a request to the credit API because
+    the parameters were not valid.
```
Minor: This docstring mentions parameters, however not all of the children of `CreditApiBadRequest` deal with parameters.

clintonb commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -29,10 +29,47 @@ class CreditProvider(TimeStampedModel):
+        unique=True,
+        help_text=ugettext_lazy(
+            "Unique identifier for this credit provider. "
+            "Only alphanumeric characters and hyphens (-) are allowed. "
```
This constraint is not enforced anywhere.

clintonb commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -29,10 +29,47 @@ class CreditProvider(TimeStampedModel):
     """
+    provider_id = models.CharField(
+        max_length=255,
+        db_index=True,
```
MySQL will create an index by default since this field is unique. There is no need for `db_index`.

clintonb commented on openedx/core/djangoapps/credit/models.py:
```diff
@@ -29,10 +29,47 @@ class CreditProvider(TimeStampedModel):
+
+    active = models.BooleanField(
+        default=True,
+        help_text=ugettext_lazy("Whether the credit provider is currently enabled.")
```
Any reason you're not aliasing `ugettext_lazy` as `_`?

wedaly replied:
I was thinking it would be confusing to alias to `_` if in the future someone tries to translate a string in an instance method (where they'd want `ugettext` instead of `ugettext_lazy`).

wedaly commented:
@clintonb Thanks for the review!  I think I've addressed all your comments.  Still not sure if the `CreditApiBadRequest` docstring is clear enough, so please let me know if there's a better way to name this.

clintonb commented on lms/urls.py:
```diff
@@ -113,6 +114,12 @@
+if settings.FEATURES.get("ENABLE_CREDIT_API"):
+    # Credit API end-points
+    urlpatterns += (
+        url(r'^api/credit/', include('openedx.core.djangoapps.credit.urls')),
```
Use a namespace: https://docs.djangoproject.com/en/1.8/topics/http/urls/#id4.

clintonb commented on openedx/core/djangoapps/credit/urls.py:
```diff
@@ -0,0 +1,22 @@
+    url(
+        r"^v1/provider/(?P<provider_id>[^/]+)/request/$",
+        create_credit_request,
+        name="credit_create_request"
```
Once namespaced, the `credit_` prefix can be removed.

clintonb commented:
:+1:

wedaly commented:
Awesome; I've updated the URL names to use the "credit" namespace and will merge once tests pass.

wedaly commented:
jenkins test this please

wedaly commented:
jenkins test this please

wedaly commented:
Test failure is unrelated; merging.
